### PR TITLE
Create fingerprint

### PR DESCRIPTION
**Pull request by pbochis**

### CODE REVIEW

File: coduno.go [hunk 1]
```diff
@@ -96,11 +104,11 @@ func setupHandler(handler Handler) http.HandlerFunc {
...
-		if session.IsNew {
-			http.Error(w, "Unauthorized session.", http.StatusUnauthorized)
-			return
-		}
+		// session, _ := cs.Get(r, sessionID)
+		// if session.IsNew {
+		// 	http.Error(w, "Unauthorized session.", http.StatusUnauthorized)
+		// 	return
+		// }
 		ctx := setupBaseHandler(w, r)
```
Comment (pbochis) [inline coduno.go]: considering the security will be changed, this makes it easier to develop locally untill that happens

File: coduno.go [hunk 2]
```diff
@@ -55,12 +56,19 @@ func main() {
 	r.HandleFunc("/api/fingerprint/company/{companyId}", setupHandler(controllers.LoadFingerprintsByCompanyID))
 	r.HandleFunc("/api/company", setupHandler(controllers.CreateCompany))
 	r.HandleFunc("/api/fingerprint", setupHandler(controllers.CreateFingerprint))
-
+	r.HandleFunc("/api/challange/company/{companyId}", setupHandler(controllers.GetChallangesForCompany))
```
Comment (flowlo) [inline coduno.go]: `s/challange/challenge/`

File: controllers/challange.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+	"github.com/coduno/app/models"
+	"github.com/coduno/app/util"
+	"github.com/gorilla/mux"
+	"golang.org/x/net/context"
+	"google.golang.org/appengine/datastore"
+)
+
+// GetChallangesForCompany -
+func GetChallangesForCompany(w http.ResponseWriter, r *http.Request, ctx context.Context) {
```
Comment (flowlo) [inline controllers/challange.go]: `s/Challange/Challenge/`

File: controllers/challange.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	var err error
+
+	if !util.CheckMethod(w, r, "GET") {
+		return
+	}
+
+	companyKey := mux.Vars(r)["companyId"]
+
+	if companyKey == "" {
+		http.Error(w, "Ned to provde a company id", http.StatusInternalServerError)
```
Comment (flowlo) [inline controllers/challange.go]: `"Need to provide ..."`

File: controllers/challange.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+		return
+	}
+
+	key, err := datastore.DecodeKey(companyKey)
+
+	if err != nil {
+		http.Error(w, "Ned to provde a company id", http.StatusInternalServerError)
```
Comment (flowlo) [inline controllers/challange.go]: `"Need to provide ..."`

File: controllers/challange.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+		return
+	}
+
+	q := datastore.NewQuery(models.ChallangeKind).Filter("Company=", key)
```
Comment (flowlo) [inline controllers/challange.go]: `s/Challange/Challenge/`

File: controllers/challange.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+	if err != nil {
+		http.Error(w, "Internal Server error: "+err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	for i := 0; i < len(keys); i++ {
+		challenges[i].EntityID = keys[i].Encode()
+	}
+
+	response["challanges"] = challenges
```
Comment (flowlo) [inline controllers/challange.go]: `s/challange/challenge/`

File: controllers/challange.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+	body, err := json.Marshal(response)
+
+	if err != nil {
+		http.Error(w, "Internal Server error: "+err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.Write([]byte(body))
```
Comment (flowlo) [inline controllers/challange.go]: No need to cast here. `json.Marshal()` returns `[]byte`.

File: controllers/challange.go [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+	if err != nil {
+		http.Error(w, "Ned to provde a company id", http.StatusInternalServerError)
+		return
+	}
+
+	q := datastore.NewQuery(models.ChallangeKind).Filter("Company=", key)
+
+	var challenges []models.Challenge
+
+	response := make(map[string]interface{})
```
Comment (flowlo) [inline controllers/challange.go]: I'd rather move this further down, to where it is used.

File: controllers/fingerprint.go [hunk 10]
```diff
@@ -14,6 +15,14 @@ import (
 	"github.com/m4rw3r/uuid"
 )
 
+//FingerprintData -
```
Comment (flowlo) [inline controllers/fingerprint.go]: Leaving a comment like this is worse than leaving no comment. Delete it or provide content.

File: controllers/fingerprint.go [hunk 11]
```diff
@@ -23,12 +32,20 @@ func CreateFingerprint(w http.ResponseWriter, r *http.Request, ctx context.Conte
...
-	lastName := mux.Vars(r)["lastName"]
-	email := mux.Vars(r)["email"]
+	var fingerprintData FingerprintData
+	if err = json.Unmarshal(body, &fingerprintData); err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
 
-	coder := models.Coder{FirstName: firstName, LastName: lastName, Email: email}
+	coder := models.Coder{FirstName: fingerprintData.FirstName, LastName: fingerprintData.LastName, Email: fingerprintData.Email}
```
Comment (flowlo) [inline controllers/fingerprint.go]: Please split this into multiple lines.

File: models/challenge.go [hunk 12]
```diff
@@ -6,7 +6,7 @@ import (
 )
 
 //ChallangeKind name of the collection in datastore
-const ChallangeKind = "challange"
+const ChallangeKind = "challenges"
```
Comment (flowlo) [inline models/challenge.go]: `s/Challange/Challenge/`

File: util/password/password.go [hunk 13]
```diff
@@ -49,7 +54,7 @@ func Hash(password []byte) (hash []byte, err error) {
 		return
 	}
 
-	return bcrypt.GenerateFromPassword(password, 0)
+	return bcrypt.GenerateFromPassword(password, Cost)
```
Comment (flowlo) [inline util/password/password.go]: I dislike this change. My intention on passing 0 here is to always use a reasonable default provided by the maintainer(s) of `golang.org/x/crypto/bcrypt`.

**Pull request merged**